src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs

```diff
@@ -151,59 +152,55 @@ public void SetErrorHandler(ConsensusRuleBase errorHandler)
         }
 
         /// <inheritdoc/>
-        public void HeaderValidation(ValidationContext validationContext)
+        public ValidationContext HeaderValidation(ChainedHeader header)
         {
-            Guard.NotNull(validationContext, nameof(validationContext));
+            Guard.NotNull(header, nameof(header));
 
+            var validationContext = new ValidationContext { ChainedHeaderToValidate = header };
             RuleContext ruleContext = this.CreateRuleContext(validationContext);
 
             this.ExecuteRules(this.headerValidationRules, ruleContext);
+
+            return validationContext;
         }
 
         /// <inheritdoc/>
-        public void IntegrityValidation(ValidationContext validationContext)
+        public ValidationContext IntegrityValidation(ChainedHeaderBlock chainedHeaderBlock)
```

Comment:
If you split this to two arguments, you will probably save a lot of allocations

Reply:
Fixed